Headers/DebugServer2/Utils/Bits.h

```diff
@@ -0,0 +1,82 @@
+//
+// Copyright (c) 2014-present, Facebook, Inc.
+// All rights reserved.
+//
+// This source code is licensed under the University of Illinois/NCSA Open
+// Source License found in the LICENSE file in the root directory of this
+// source tree. An additional grant of patent rights can be found in the
+// PATENTS file in the same directory.
+//
+
+#ifndef __DebugServer2_Utils_Bits_h
+#define __DebugServer2_Utils_Bits_h
+
+#include "DebugServer2/Utils/CompilerSupport.h"
+
+namespace ds2 {
+namespace Utils {
+namespace {
+// We apply builtin bit functions on chunks of sizeof(int) * 4 bits.
+const size_t kChunkSize = sizeof(int) * 4;
+}
+
+static inline unsigned int FFS(unsigned int number) {
+#if __has_builtin(__builtin_ffs)
+  return __builtin_ffs(number);
+#elif defined(_MSC_VER)
+  unsigned long index;
+  if (_BitScanForward(&index, number)) {
+    return index + 1;
+  }
+  return 0;
+#else
+  for (int i = 1; number != 0; i++, number >>= 1) {
+    if (number & 1) {
+      return i;
+    }
+  }
+  return 0;
+#endif
+}
+
+template <typename T> static inline unsigned int FFS(T number) {
+  while (number != 0) {
+    int ffs = FFS(number);
+    if (ffs != 0) {
+      return ffs;
+    }
+    number >>= kChunkSize;
+  }
+  return 0;
+}
+
+static inline unsigned int PopCount(unsigned int number) {
+  int count = 0;
+#if __has_builtin(__builtin_popcount)
+  while (number != 0) {
+    count += __builtin_popcount(number);
+    number >>= kChunkSize;
+  }
+#elif defined(_MSC_VER)
+  count = _CountOneBits(number);
+#else
+  while (number != 0) {
+    count++;
+    number = number & (number - 1);
+  }
+#endif
+  return count;
+}
+
+template <typename T> static inline unsigned int PopCount(T number) {
+  int count = 0;
+  while (number != 0) {
+    count += PopCount(number);
+    number >>= kChunkSize;
```

Comment:
Are you sure `std::bitset` is as performant as the implementation @a20012251
<https://github.com/a20012251> wrote?

On Tue, Mar 8, 2016 at 4:28 PM, Dave Lee <notifications@github.com> wrote:

> In Headers/DebugServer2/Utils/Bits.h
> <https://github.com/facebook/ds2/pull/211#discussion_r55453958>:
>
> > +#elif defined(_MSC_VER)
> > +  count = _CountOneBits(number);
> > +#else
> > +  while (number != 0) {
> > +    count++;
> > +    number = number & (number - 1);
> > +  }
> > +#endif
> > +  return count;
> > +}
> > +
> > +template <typename T> static inline unsigned int PopCount(T number) {
> > +  int count = 0;
> > +  while (number != 0) {
> > +    count += PopCount(number);
> > +    number >>= kChunkSize;
>
> This implementation could be:
>
> std::bitset<sizeof(T) * 8> x(number);return x.count();
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/ds2/pull/211/files#r55453958>.
>



-- 
Stephane Sezer